### PR TITLE
kernel: utilities: add crc32b implementation

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

Alternative to #3808.

This is helpful for, for one example, computing ShortIDs from process names.

The `Compress` trait has one function:

```rust
fn to_short_id(&self, process: &dyn Process, credentials: &TbfFooterV2Credentials) -> ShortID;
```

I want to add a function with the signature `fn (&'static str) -> u32)` so I can implement `Compress` and assign ShortIDs based on process names. 

This complements #3818, which is an app checker that looks like:

```rust
pub struct AppCheckerNames<'a, F: Fn(&'static str) -> u32> {
    hasher: &'a F,
    ...
}
```

For my purposes (and for most simple use cases where apps are identified by their name), it's not important _what_ the hash function is, as long as it is reasonably ok at generating different short IDs for different strings.

Now, I've been convinced to change the signature of the function in this PR to `fn (&[u8]) -> u32`. That makes this a little more general. For my use case then I will create a small wrapper function in the board main.rs, or do something like `&|s| crc32b(s.as_bytes())`.

The other reason I want _any_ simple function like `fn (&'static str) -> u32)` is it makes actually using AppID in board files fairly ergonomic. So If I have a capsule and I want to give some resource to a specific app, I can do something like:

```rust
my_capsule::new(
  // Identify the app that is allowed to use this capsule
  ShortID::Fixed(NonZeroU32::new(kernel::utilities::helpers::crc("circle")).unwrap())
  // Other resources the capsule needs
  i2c, ...
)
```

I guess now I have to do:

```
fn my_hash(s: &'static str) -> u32 {
  kernel::utilities::helpers::crc(s.as_bytes())
}

my_capsule::new(
  // Identify the app that is allowed to use this capsule
  ShortID::Fixed(NonZeroU32::new(my_hash("circle")).unwrap())
  // Other resources the capsule needs
  i2c, ...
)
```

but you get the point.

What makes this nice is that it's easy for someone to know the ShortID for an app since we already use process names. Changing which app gets a resource is just updating the human readable string.

If we want to make AppID usable, having ways to generate ShortIDs that is intuitive is very helpful.

### Testing Strategy

Comparing to online calculator.

### TODO or Help Wanted

n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: kernel/src/utilities/helpers.rs [hunk 1]
```diff
@@ -42,3 +42,25 @@ macro_rules! count_expressions {
     ($head:expr $(,)?) => (1usize);
     ($head:expr, $($tail:expr),* $(,)?) => (1usize + count_expressions!($($tail),*));
 }
+
+/// Compute the CRC-32B checksum of a string.
+///
+/// Implementation based on https://lxp32.github.io/docs/a-simple-example-crc32-calculation/.
+/// Online calculator here: https://md5calc.com/hash/crc32b
+pub fn crc32b_str(s: &'static str) -> u32 {
```
Comment (lschuermann) [inline kernel/src/utilities/helpers.rs:50]: Might be nicer to have this accept a byte slice instead. This way we can use [`core:;str::as_bytes`](https://doc.rust-lang.org/stable/core/primitive.str.html#method.as_bytes) to calculate a CRC over a string, but also support other non-unicode byte sequences. Shouldn't need a lot of changes, just iterating over the bytes and removing the cast.

**Review by lschuermann — CHANGES_REQUESTED** [review]

**Comment by alistair23** [comment]
Why not use the existing [CRC implementation](https://github.com/tock/tock/blob/master/libraries/tickv/src/crc32.rs)?

**Comment by bradjc** [comment]
> Why not use the existing [CRC implementation](https://github.com/tock/tock/blob/master/libraries/tickv/src/crc32.rs)?

I think adding a dependency of tickv for this use case is pretty heavyweight. Also the kernel won't have tickv as a dependency and I need to use this in the kernel. I don't actually need CRC, I just need something of `fn(str) -> u32` that I can get merged into Tock.

**Comment by alistair23** [comment]
You could copy the same implementation though. Keep all our CRCs the same

**Review by lschuermann — APPROVED** [review]

**Comment by bradjc** [comment]
> You could copy the same implementation though. Keep all our CRCs the same

I need something so I can do something like:

```rust
ShortID::Fixed(NonZeroU32::new(kernel::utilities::helpers::crc("circle")).unwrap())
```

Any extra machinery would just complicate that code for I don't think much benefit. 

Use case: https://github.com/tock/tock/pull/3817/files#diff-98d9b86eaf1d01d892fb1c96c448e2799cf55fa442b5ff12aa4b5d85c3d16de5R506

**Comment by alistair23** [comment]
I don't follow. You said in the PR description that this is helpful for computing shortIDs. Why do we need to use CRC-32B? Could we use CRC-32 to try to keep all of our CRCs the same?

Maybe it's worth saying why we need this exact implementation in the PR description? It's not really clear

**Comment by bradjc** [comment]
Version that matches tickv in #3829

**Comment by bradjc** [comment]
I don't think anyone wants to add crc32b. But the PR description is still helpful.

**Pull request closed without merging**